### ISSUE TITLE
Rethink the file object

### ISSUE THREAD

**Issue by sindresorhus** [wooorm/vfile#8]
There are some weirdness in the API. Like, why is `basename()` a method, while `extension` is a property?

I think all the of the properties should be dynamic getters, so if you assign to for example `extension`, accessing `filename` should reflect the change.

It's missing stuff like `.stat`, `.path` (full path), [`.stem`](https://github.com/gulpjs/vinyl#stem),

**Comment by wooorm** [comment]
Yes, definitely!

From the previous issue:

> There's benefit in being able to detect that it's a VFile.

Sure, but that could be done by checking `'filePath' in thing && 'messages' in thing`, and the like.

> It's missing stuff like .stat, .path (full path), .stem,

Or, rather, remove all those fs/path related methods and just use `fs.stat`, `path.basename`, `path.extname`, and the like directly?

> I think all the of the properties should be dynamic getters, so if you assign to for example extension, accessing filename should reflect the change.

TBH I prefer not to use dark magic 😬 😢, instead, why not use a module to change, for example, the extname in a path?

**Comment by sindresorhus** [comment]
> Sure, but that could be done by checking 'filePath' in thing && 'messages' in thing, and the like.

Any object could have those properties. That would make it very likely to have false-positives.

> Or, rather, remove all those fs/path related methods and just use fs.stat, path.basename, path.extname, and the like directly?

Uhm, it's a virtual file. That would miss the whole point of it. The path might not even be an actual existing file, but could be a future existing file or the file could be in a different location on disk.

> TBH I prefer not to use dark magic 😬 😢, instead, why not use a module to change, for example, the extname in a path?

I don't see how that is dark magic. It's actually expected behavior for me and previously many users on the `vinyl` issue tracker. It's how `vinyl` works now btw. Think of the `path` as an internal model and those properties changes that internal model and access it.

**Comment by wooorm** [comment]
> Any object could have those properties. That would make it very likely to have false-positives.

If it looks like a vfile, and warns like a vfile? I kinda like the looseness and non-typed nature of JS, also here.

> Uhm, it's a virtual file. That would miss the whole point of it. The path might not even be an actual existing file, but could be a future existing file or the file could be in a different location on disk.

Yeah but there’s no reason in duplicating vinyl, or path, or fs in the core, whether its a virtual file or not, `require('path').basename(vfile.filePath)` (assuming we’re moving to `string`) would work.

###### Work in Progress

```diff
  `VFile#contents`

- `VFile#directory`

- `VFile#filename`

- `VFile#extension`

- `VFile#basename()`

  `VFile#quiet`
  (this property is a bit like `dangerouslySetInnerHTML`, in that it
  should only be used if you know exactly what you’re doing, I still
  think its required, and should be opt-in though);

  `VFile#messages`

- `VFile#history`
  (don’t think it’s needed to keep track of the whole history, maybe
  an `origin` for the first location upon reading, if ever, could work?)

  `VFile#toString()
  (needed, buffers will be added too)

- `VFile#filePath()`

+ `VFile#filePath`

- `VFile#move(options)`

- `VFile#namespace(key)`
  (can be done by standardising custom properties, e.g., `origin`
  mentioned above in `VFile#history`);

  `VFile#message(reason[, position[, ruleId]])`

- `VFile#warn(reason[, position[, ruleId]])`
  Remove `VFile#message` and move `VFile#warn` to its name instead;

- `VFile#fail(reason[, position[, ruleId]])`
  Rename to `VFile#throw`, albeit with the `#warn` > `#message`, `fail`
  may be sufficiently different;

- `VFile#hasFailed()`
  Remove. Either replace by a module `vfile-count` or something, or
  keep track of number of the count of `failures` and `messages` messages.
  Related to GH-7.

  `VFileMessage`

  ...

+ `VFileMessage#source`
  Either standardise it, or standardise custom properties;

  `VFileMessage#fatal`
  Remove difference between `fatal: false` and `fatal: null`, there’s
  messages (previous warnings), and there’s fatal errors.

  ...

**Comment by wooorm** [comment]
@sindresorhus I wrote an initial proposal, GH-10, I’d love to hear your thoughts (there’s some specific questions I’m unsure about at the bottom).

**Comment by wooorm** [comment]
Closed by 1959856! 🎉
